File: Algorithms/Python/cpu_scheduling/round_robin.py
from __future__ import annotations

from queue import Queue


class Process:
    def __init__(self, process_id: int, /, *, arrival_time: int, burst_time: int):
        self.process_id = process_id
        self.arrival_time = arrival_time
        self.burst_time = burst_time
        self.remaining_time = burst_time
        self.waiting_time = 0


class GanttChart:
    def __init__(self) -> None:
        self.chart = []

    def add(self, process_id: int, /, *, time: int) -> None:
        self.chart.append((process_id, time))

    def __str__(self) -> str:
        return " -> ".join(f"P{process_id} [{time}]" for process_id, time in self.chart)


def round_robin(processes: list[Process], quantum: int) -> None:
    processes.sort(key=lambda process: process.arrival_time)
    time = 0
    queue: Queue[Process] = Queue()
    gantt_chart = GanttChart()

    avg_waiting_time = 0
    avg_turnaround_time = 0

    i = 0
    while i < len(processes) or not queue.empty():
        while i < len(processes) and processes[i].arrival_time <= time:
            queue.put(processes[i])
            i += 1

        if queue.empty():
            time += 1
            continue

        process = queue.get()
        gantt_chart.add(process.process_id, time=time)

        if process.remaining_time > quantum:
            time += quantum
            process.remaining_time -= quantum
            queue.put(process)
        else:
            time += process.remaining_time
            process.waiting_time = time - process.arrival_time - process.burst_time
            avg_waiting_time += process.waiting_time
            avg_turnaround_time += time - process.arrival_time
            process.remaining_time = 0

    avg_waiting_time /= len(processes)
    avg_turnaround_time /= len(processes)

    print("Gantt Chart:", gantt_chart)

    print("Process ID\tArrival Time\tBurst Time\tWaiting Time\tTurnaround Time")
    for process in processes:
        print(
            f"{process.process_id}\t\t{process.arrival_time}\t\t"
            f"{process.burst_time}\t\t{process.waiting_time}\t\t"
            f"{process.waiting_time + process.burst_time}",
        )
    print(f"Avg Waiting Time: {avg_waiting_time}")
    print(f"Avg Turnaround Time: {avg_turnaround_time}")


if __name__ == "__main__":
    processes = [
        Process(1, arrival_time=0, burst_time=4),
        Process(2, arrival_time=1, burst_time=3),
        Process(3, arrival_time=2, burst_time=5),
        Process(4, arrival_time=3, burst_time=2),
    ]
    round_robin(processes, quantum=2)

File: Algorithms/C/cpu_scheduling/round_robin.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

struct process
{
    int process_id;
    int arrival_time;
    int burst_time;

    // Variables to be used for calculations
    int remaining_time;
    int waiting_time;
};

struct gantt_chart_component
{
    int process_id;
    int time;
};

struct gantt_chart_component gantt_chart[100];
int gantt_chart_size = 0;

void sort_processes(struct process *processes, int num_processes);

int partition(struct process *processes, int low, int high)
{
    int pivot = processes[high].arrival_time;
    int i = low - 1;

    for (int j = low; j < high; j++)
    {
        if (processes[j].arrival_time < pivot)
        {
            i++;
            struct process temp = processes[i];
            processes[i] = processes[j];
            processes[j] = temp;
        }
    }

    struct process temp = processes[i + 1];
    processes[i + 1] = processes[high];
    processes[high] = temp;

    return i + 1;
}

void quick_sort(struct process *processes, int low, int high)
{
    if (low < high)
    {
        int pi = partition(processes, low, high);

        quick_sort(processes, low, pi - 1);
        quick_sort(processes, pi + 1, high);
    }
}

void sort_processes(struct process *processes, int num_processes)
{
    quick_sort(processes, 0, num_processes - 1);
}

/* Queue Implementation */

struct process *circular_queue[100];
int queue_size;
int front = 0;
int rear = 0;

void put(struct process *p)
{
    circular_queue[rear] = p;
    rear = (rear + 1) % queue_size;
}

struct process *get()
{
    struct process *p = circular_queue[front];
    front = (front + 1) % queue_size;
    return p;
}

bool empty()
{
    return front == rear;
}

bool full()
{
    return (rear + 1) % queue_size == front;
}

/* Round Robin Scheduling Algorithm */

void round_robin(struct process *processes, int total_processes, int time_quantum)
{
    sort_processes(processes, total_processes);
    int time = 0;

    float avg_waiting_time = 0;
    float avg_turnaround_time = 0;

    int i = 0;

    while (i < total_processes || !empty())
    {
        while (i < total_processes && processes[i].arrival_time <= time)
        {
            put(&processes[i]);
            i++;
        }

        if (empty())
        {
            time++;
            continue;
        }

        struct process *current_process = get();
        gantt_chart[gantt_chart_size].process_id = current_process->process_id;
        gantt_chart[gantt_chart_size].time = time;
        gantt_chart_size++;

        if (current_process->remaining_time > time_quantum)
        {
            time += time_quantum;
            current_process->remaining_time -= time_quantum;
            put(current_process);
        }
        else
        {
            time += current_process->remaining_time;
            current_process->waiting_time = time - current_process->arrival_time - current_process->burst_time;
            avg_waiting_time += (float)current_process->waiting_time;
            avg_turnaround_time += (float)time - (float)current_process->arrival_time;
            current_process->remaining_time = 0;
        }
    }

    avg_waiting_time /= (float)total_processes;
    avg_turnaround_time /= (float)total_processes;

    printf("Gantt Chart: ");
    for (int i = 0; i < gantt_chart_size; i++)
    {
        printf("P%d [%d] -> ", gantt_chart[i].process_id, gantt_chart[i].time);
    }
    printf("NULL\n");

    printf("Process ID\tArrival Time\tBurst Time\tWaiting Time\tTurnaround Time\n");
    for (int i = 0; i < total_processes; i++)
    {
        printf("%d\t\t%d\t\t%d\t\t%d\t\t%d\n", processes[i].process_id, processes[i].arrival_time,
               processes[i].burst_time, processes[i].waiting_time, processes[i].waiting_time + processes[i].burst_time);
    }

    printf("Average Waiting Time: %.2f\n", avg_waiting_time);
    printf("Average Turnaround Time: %.2f\n", avg_turnaround_time);
}

int main()
{
    /*
    printf("Enter the number of processes: ");
    scanf("%d", &num_processes);

    printf("Enter the time quantum: ");
    scanf("%d", &time_quantum);

    for (int i = 0; i < num_processes; i++)
    {
        printf("Enter the arrival time and burst time of process %d: ", i + 1);
        scanf("%d %d", &processes[i].arrival_time, &processes[i].burst_time);
        processes[i].process_id = i + 1;
        processes[i].remaining_time = processes[i].burst_time;
    }
    */

    int num_processes = 4;
    queue_size = num_processes;

    int time_quantum = 2;
    
    struct process p1 = {1, 0, 4, 4, 0};
    struct process p2 = {2, 1, 3, 3, 0};
    struct process p3 = {3, 2, 5, 5, 0};
    struct process p4 = {4, 3, 2, 2, 0};
    
    struct process processes[] = {p1, p2, p3, p4};

    round_robin(processes, num_processes, time_quantum);

    return 0;
}
